File: tensorflow/detector.py
# TensorFlow and tf.keras
import tensorflow as tf
from tensorflow import keras

# Helper libraries
import numpy as np
import matplotlib.pyplot as plt
import os
import json
import h5py

class Detector():
    CLASS_NAMES = ['OK', 'NOTOK']
    INPUT_SHAPE = (17, 2)
    LOSS = 'sparse_categorical_crossentropy'
    MODEL = None
    METRICS = ['accuracy']
    TRAIN_COEF = 0.1

    def __init__(self, class_names, input_shape):
        self.INPUT_SHAPE = input_shape
        self.CLASS_NAMES = class_names
        label_layers = len(self.CLASS_NAMES)
        self.MODEL = keras.Sequential([
            keras.layers.Flatten(input_shape=input_shape),
            keras.layers.Dense(128, activation=tf.nn.relu),
            keras.layers.Dense(label_layers, activation=tf.nn.softmax)
        ])
        self.MODEL.compile(optimizer=tf.train.AdamOptimizer(), 
              loss=self.LOSS,
              metrics=self.METRICS)

    def train_from_file(self,pathfile, epohs=5):
        data = {"input": [], "output":[]}

        with open(pathfile, "r") as json_data:
            data = json.load(json_data)


        self.train(data)

    def train(self,indata, epohs=5):
        rand = np.random.choice(np.arange(0, int(len(indata))), size=int(len(indata)*self.TRAIN_COEF), replace=False)
        i = 0
        train_input = []
        train_labels = []
        test_input = []
        test_labels = []
        for item in indata:
            if i not in rand:
                train_input.append(item['input'])
                train_labels.append(item['output'])
            else:
                test_input.append(item['input'])
                test_labels.append(item['output'])
            i += 1
        self.MODEL.fit(np.array(train_input), train_labels, epochs=epohs)
        test_loss, test_acc = self.MODEL.evaluate(np.array(test_input), np.array(test_labels))
        print('Test accuracy:', test_acc)

    def detect(self, input):
        predictions = self.MODEL.predict([input])
        return np.argmax(predictions[0])

    def save(self):
        model_json = self.MODEL.to_json()
        with open("model.json", "w") as json_file:
            json_file.write(model_json)
        self.MODEL.save_weights("./checkpoints")

File: tensorflowjs/detector.js
require('@tensorflow/tfjs-node');
const tf = require('@tensorflow/tfjs');
const _ = require('lodash');
const data = require('../dataset');

class Detector {

  constructor({ classNames, inputShape }) {
    this.MODEL = tf.sequential()
    this.MODEL.add(tf.layers.flatten({inputShape: inputShape}));
    this.MODEL.add(tf.layers.dense({ units: 128, activation: 'relu' }));
    this.MODEL.add(tf.layers.dense({ units: 1, activation: 'softmax' }));
    this.MODEL.compile({optimizer: 'adam', loss: 'categoricalCrossentropy', metrics: ['accuracy']})
  }
  async train(data, epochs = 5) {
    const randomIndexes = choice(data, data.length * 0.1);
    const trainInput = [];
    const trainLabels = [];
    const testInput = [];
    const testLabels = [];
    data.forEach(({ input, output }, i) => {
      if (randomIndexes.includes(i)) {
        testInput.push(input);
        testLabels.push(output);
        return;
      }

      trainInput.push(input);
      trainLabels.push(output);
    });
    for (let i = 0; i < trainInput.length; i++) {
      const trainTensor = tf.tensor(trainInput[i]);
      const trainAnswerTensor = tf.tensor(trainLabels[i]);
      console.log(trainTensor,trainAnswerTensor)
      await this.MODEL.fit(trainTensor, trainAnswerTensor, { epochs }).catch(e => console.log(e));
    }
    // const trainTensor = tf.tensor(trainInput);
    // const trainAnswerTensor = tf.tensor(trainLabels.map(n => [n, 0]))
    // console.log(trainInput.length,trainTensor,trainAnswerTensor)
    // await this.MODEL.fit(trainTensor, trainAnswerTensor, { epochs }).catch(e => console.log(e));
    // const testInputTensor = tf.tensor(trainInput);
    // const testLabelsTensor = tf.tensor(trainLabels.map(n => [n, 0]))
    // const result = this.MODEL.evaluate(testInputTensor, testLabelsTensor)
    // console.log(result[1].print())
  }

  predict({input, output}) {
    const res = this.MODEL.predict(tf.tensor([input]));
    console.log(res.print(), output)
  }
}

function choice (data, count) {
  const values = [];
  do {
    const index = _.random(0, data.length);
    if(!values.includes(index)) values.push(index);
  } while (values.length !== count);
  return values;
}

const d = new Detector({ classNames: ["not ok", "ok"], inputShape: [17, 2] });
d.train(data, 5)//.then(() => {
//   d.predict(data.filter(({output}) => !output)[8])
// });
//
